test/e2e/catalog_e2e_test.go

```diff
@@ -761,3 +767,98 @@ func replicateCatalogPod(t *testing.T, c operatorclient.ClientInterface, crc ver
 
 	return copied
 }
+
+func TestGrpcConnectionStateIsNeverTransient(t *testing.T) {
+	// Create gRPC CatalogSource using an external registry image (community-operators)
+	// Wait for a registry pod to be created
+	// Monitor status and report failure if transitions do not progress cleanly (connecting -> connected)
+	// Delete the registry pod (part of cleaning up)
+	defer cleaner.NotifyTestComplete(t, true)
+	sourceName := genName("catalog-")
+
+	// Create gRPC CatalogSource using an external registry image (community-operators)
+	source := &v1alpha1.CatalogSource{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1alpha1.CatalogSourceKind,
+			APIVersion: v1alpha1.CatalogSourceCRDAPIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sourceName,
+			Namespace: testNamespace,
+			Labels:    map[string]string{reconciler.CatalogSourceLabelKey: sourceName},
+		},
+		Spec: v1alpha1.CatalogSourceSpec{
+			SourceType: v1alpha1.SourceTypeGrpc,
+			Image:      communityOperatorsImage,
+		},
+	}
+
+	crc := newCRClient(t)
+	selector := labels.SelectorFromSet(map[string]string{reconciler.CatalogSourceLabelKey: sourceName})
+	catSrcWatcher, err := crc.OperatorsV1alpha1().CatalogSources(testNamespace).Watch(metav1.ListOptions{LabelSelector: selector.String()})
+	require.NoError(t, err)
+
+	// Monitor GRPC connection status updates
+	ctx, cancel := context.WithCancel(context.Background())
+	// cancel called in defer func below
+	done := make(chan struct{})
+	errExit := make(chan error)
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case evt, ok := <-catSrcWatcher.ResultChan():
+				if !ok {
+					errExit <- errors.New("watch channel closed unexpectedly")
+					return
+				}
+				if evt.Type == watch.Modified {
+					catSrc, ok := evt.Object.(*v1alpha1.CatalogSource)
+					if !ok {
+						errExit <- errors.New("watch returned unexpected object type")
+						return
+					}
+					t.Logf("connectionState=%v", catSrc.Status.GRPCConnectionState)
+					if catSrc.Status.GRPCConnectionState != nil {
+						if catSrc.Status.GRPCConnectionState.LastObservedState == connectivity.Ready.String() {
+							done <- struct{}{}
+						}
+						require.NotEqual(t, connectivity.TransientFailure.String(), catSrc.Status.GRPCConnectionState.LastObservedState)
+					}
+				}
+
+			}
+		}
```

Comment:
Since the watch was started earler, I think this can be merged with the `for {...}` at the bottom of this test.

Reply:
I don't want to miss any state changes, so I really want to start monitoring the watch before the catalog source is created. Since the result chan is unbuffered, I think that's the only safe way to do it, right?